library/src/main/java/com/google/android/exoplayer/hls/parser/MpaReader.java

```diff
@@ -0,0 +1,297 @@
+     * that the data should be written into {@code target} starting from an offset of zero.
+     *
+     * @param source The source from which to read.
+     * @param target The target into which data is to be read.
+     * @param targetLength The target length of the read.
+     * @return Whether the target length was reached.
+     */
+    private boolean continueRead(ParsableByteArray source, byte[] target, int targetLength) {
+        int bytesToRead = Math.min(source.bytesLeft(), targetLength - bytesRead);
+        source.readBytes(target, bytesRead, bytesToRead);
+        bytesRead += bytesToRead;
+        return bytesRead == targetLength;
+    }
+
+    /**
+     * Locates the next sync word, advancing the position to the byte that immediately follows it.
+     * If a sync word was not located, the position is advanced to the limit.
+     *
+     * @param pesBuffer The buffer whose position should be advanced.
+     * @return True if a sync word position was found. False otherwise.
+     */
+    private boolean skipToNextSync(ParsableByteArray pesBuffer) {
+        byte[] mpaData = pesBuffer.data;
+        int startOffset = pesBuffer.getPosition();
+        int endOffset = pesBuffer.limit();
+        for (int i = startOffset; i < endOffset - 1; i++) {
+            int syncBits = (mpaData[i] << 8 ) | mpaData[i + 1];
+            if ((syncBits & 0xFFF0) == 0xFFF0) {
+                hasCrc = (mpaData[i + 1] & 0x1) == 0;
+                pesBuffer.setPosition(i);
+                return true;
+            }
+        }
+        pesBuffer.setPosition(endOffset);
+        return false;
+    }
+
+    /**
+     * Calculates MPEG Audio frame size
+     *
+     * @param layer The MPEG layer
+     * @param LSF Low Sample rate Format (MPEG 2)
+     * @param bitrate The bitrate in bits per second
+     * @param samplesPerSec The sampling rate in hertz
+     * @param paddingSize
+     * @return Frame size in bytes
+     */
+    private static int CalcMpaFrameSize (int layer, int LSF, int bitrate, int samplesPerSec, int paddingSize) {
+        return (int)(Math.floor(MPA_COEFFICIENTS[LSF][layer] * bitrate / samplesPerSec) + paddingSize) * MPA_SLOT_SIZE[layer];
+    }
+
+    /**
+     * Parses the sample header.
+     */
+    private void parseHeader() {
+        mpaScratch.setPosition(0);
+
+        if (!hasMediaFormat()) {
+            mpaScratch.skipBits(12);
+            int isLSF = (!mpaScratch.readBit()) ? 1 : 0;
+            int layer = mpaScratch.readBits(2) ^ 3;
+            mpaScratch.skipBits(1);
+            int audioObjectType = 32 + layer;
+            int bitRate = MPA_BITRATES[isLSF][layer][mpaScratch.readBits(4)];
+            int sampleRate = MPA_SAMPLING_RATES[3 - isLSF][mpaScratch.readBits(2)];
+            int sampleRateIndex = 0;
+            for (; sampleRateIndex < (CodecSpecificDataUtil.AUDIO_SPECIFIC_CONFIG_SAMPLING_RATE_TABLE).length; sampleRateIndex++) {
```

Comment:
I was referring to this method https://github.com/google/ExoPlayer/blob/master/library/src/main/java/com/google/android/exoplayer/util/CodecSpecificDataUtil.java#L84
It's better to leave them private and expose a getSampleRateIndex(int sampleRate) function CodecSpecificDataUtil, and reuse it here an in the method I referred to. Same for channelConfig.